go/fury/meta/meta_string_decoder.go

```diff
@@ -0,0 +1,143 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package meta
+
+import (
+	"fmt"
+	"strings"
+)
+
+type Decoder struct {
+	specialChar1 byte
+	specialChar2 byte
+}
+
+func NewDecoder(specialCh1 byte, specialCh2 byte) *Decoder {
+	return &Decoder{
+		specialChar1: specialCh1,
+		specialChar2: specialCh2,
+	}
+}
+
+// Decode
+// Accept an encoded byte array, and the encoding method
+// We also need to accept numBits, because we don't know how many characters are in the byte array,
+// and numBits are the actual valid bits
+func (d *Decoder) Decode(data []byte, encoding Encoding, numBits int) string {
+	switch encoding {
+	case LOWER_SPECIAL:
+		return d.decodeGeneric(data, encoding, numBits)
+	case LOWER_UPPER_DIGIT_SPECIAL:
+		return d.decodeGeneric(data, encoding, numBits)
+	case FIRST_TO_LOWER_SPECIAL:
+		return strings.ToTitle(d.decodeGeneric(data, LOWER_SPECIAL, numBits))
+	case ALL_TO_LOWER_SPECIAL:
+		return d.decodeRepAllToLowerSpecial(data, LOWER_SPECIAL, numBits)
+	case UTF_8:
+		return string(data)
+	default:
+		panic("Unexpected encoding flag: {encoding}")
+	}
+}
+
+// DecodeGeneric
+// algorithm is LowerSpecial or LowerUpperDigit
+func (d *Decoder) decodeGeneric(data []byte, algorithm Encoding, numBits int) string {
+	bitsPerChar := 5
+	if algorithm == LOWER_UPPER_DIGIT_SPECIAL {
+		bitsPerChar = 6
+	}
+	// Retrieve 5 bits every iteration from data, convert them to characters, and save them to chars
+	// "abc" encoded as [00000] [000,01] [00010] [0, corresponding to three bytes, which are 0, 68, 0 (68 = 64 + 4)
+	// In order, take the highest digit first, then the lower
+	chars := make([]byte, 0)
```

Comment:
Hi @qingoba , first of all thanks for your contribution to Fury.

Could we specify the capacity when initializing `chars` to avoid the performance degradation caused by slice expansion?